### PR TITLE
CSP-1314: GetProviderLegalEntities endpoint migration

### PR DESCRIPTION
**Pull request by shaun-downey-education**
- Migrated the GetProviderLegalEntities endpoint from the **SFA.DAS.ProviderRelationships** to the **SFA.DAS.PR.Api** application.
- Added unit test coverage for all related classes.

### CODE REVIEW

File: src/SFA.DAS.PR.Data/Repositories/AccountProviderLegalEntitiesReadRepository.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+{
+    public Task<List<AccountProviderLegalEntity>> GetAccountProviderLegalEntities(long? ukprn, string? accountHashId, List<Operation> operations, CancellationToken cancellationToken)
+    {
+        return _providerRelationshipsDataContext.AccountProviderLegalEntities
+            .Include(a => a.AccountProvider)
+                .ThenInclude(a => a.Account)
+            .Include(a => a.Permissions)
+            .Include(a => a.AccountLegalEntity)
+            .Where(t =>
+                (!ukprn.HasValue || t.AccountProvider.ProviderUkprn == ukprn) &&
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Data/Repositories/AccountProviderLegalEntitiesReadRepository.cs:16]: `!ukprn.HasValue`  is not going to work out in the query vey well.

Comment (DevenKShah) [reply src/SFA.DAS.PR.Data/Repositories/AccountProviderLegalEntitiesReadRepository.cs:16]: I see there is validation for Ukprn, so in the repository, the parameter should not be nullable for ukprn and account hash id

File: src/SFA.DAS.PR.Application/AccountProviderLegalEntities/Queries/GetAccountProviderLegalEntities/AccountProviderLegalEntityModel.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+﻿using SFA.DAS.PR.Domain.Entities;
+
+namespace SFA.DAS.PR.Application.AccountProviderLegalEntities.Queries.GetAccountProviderLegalEntities;
+
+public class AccountProviderLegalEntityModel
+{
+    public long AccountId {  get; set; }
+    public string? AccountHashedId { get; set; }
+    public string? AccountPublicHashedId { get; set; }
+    public string? AccountName { get; set; }
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Application/AccountProviderLegalEntities/Queries/GetAccountProviderLegalEntities/AccountProviderLegalEntityModel.cs:10]: Why are these nullable?

File: src/SFA.DAS.PR.Application/AccountProviderLegalEntities/Queries/GetAccountProviderLegalEntities/AccountProviderLegalEntityModel.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    public long AccountLegalEntityId { get; set; }
+    public string? AccountLegalEntityPublicHashedId { get; set; }
+    public string? AccountLegalEntityName { get; set; }
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Application/AccountProviderLegalEntities/Queries/GetAccountProviderLegalEntities/AccountProviderLegalEntityModel.cs:13]: these should be non-nullable right?

File: src/SFA.DAS.PR.Application/AccountProviderLegalEntities/Queries/GetAccountProviderLegalEntities/GetAccountProviderLegalEntitiesQueryValidator.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+public class GetAccountProviderLegalEntitiesQueryValidator : AbstractValidator<GetAccountProviderLegalEntitiesQuery>
+{
+    public const string UkprnAccountHashIdValidationMessage = "Either one of Ukprn or AccountHashId is required.";
+    public const string UkprnFormatValidationMessage = "Currently a Ukprn must start with the value 1 and should be 8 digits long.";
+    public const string OperationFilterValidationMessage = "Currently at least one operation filter must to be supplied.";
+    public const string OperationFilterFormatValidationMessage = "Operation filter values are limited to 0, 1 or 2.";
+    public GetAccountProviderLegalEntitiesQueryValidator()
+    {
+        RuleFor(x => x.Ukprn)
+            .Must(Ukprn => Ukprn.HasValue)
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Application/AccountProviderLegalEntities/Queries/GetAccountProviderLegalEntities/GetAccountProviderLegalEntitiesQueryValidator.cs:13]: We used NotEmpty before, is that not an option any more, this should check for both null and zero so there is no need for a `When` condition

File: src/SFA.DAS.PR.Application/AccountProviderLegalEntities/Queries/GetAccountProviderLegalEntities/GetAccountProviderLegalEntitiesQueryValidator.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            .WithMessage(UkprnAccountHashIdValidationMessage)
+            .When(x => string.IsNullOrEmpty(x.AccountHashedId));
+
+        RuleFor(x => x.AccountHashedId)
+            .Must(AccountHashedId => !string.IsNullOrWhiteSpace(AccountHashedId))
+            .WithMessage(UkprnAccountHashIdValidationMessage)
+            .When(x => !x.Ukprn.HasValue);
+
+        RuleFor(model => model.Ukprn)
```
Comment (DevenKShah) [inline src/SFA.DAS.PR.Application/AccountProviderLegalEntities/Queries/GetAccountProviderLegalEntities/GetAccountProviderLegalEntitiesQueryValidator.cs:22]: This should be chained with NotEmpty if possible to keep things together. Also we need to start looking at reusability of these validations.

File: src/SFA.DAS.PR.Data/Repositories/AccountProviderLegalEntitiesReadRepository.cs (hunk 1 above)
Comment (shaun-downey-education) [reply src/SFA.DAS.PR.Data/Repositories/AccountProviderLegalEntitiesReadRepository.cs:16]: The logic specifies that either the Ukprn or the PubicHashId must be supplied so either or can be nullable according to 1314

File: src/SFA.DAS.PR.Application/AccountProviderLegalEntities/Queries/GetAccountProviderLegalEntities/GetAccountProviderLegalEntitiesQueryValidator.cs (hunk 5 above)
Comment (shaun-downey-education) [reply src/SFA.DAS.PR.Application/AccountProviderLegalEntities/Queries/GetAccountProviderLegalEntities/GetAccountProviderLegalEntitiesQueryValidator.cs:22]: Spoke to Mark about the reusability, as either Ukprn or PublicHashId can be null for this specific endpoint the validators cannot be reused for his specific scenario.

File: src/SFA.DAS.PR.Application/AccountProviderLegalEntities/Queries/GetAccountProviderLegalEntities/GetAccountProviderLegalEntitiesQueryValidator.cs (hunk 4 above)
Comment (shaun-downey-education) [reply src/SFA.DAS.PR.Application/AccountProviderLegalEntities/Queries/GetAccountProviderLegalEntities/GetAccountProviderLegalEntitiesQueryValidator.cs:13]: @DevenKShah The when in this instance is used to validate the opposite parameter in the request. Ukprn must have a value when AccountHashId is empty and vice versa.

**Review by MarkFCain — APPROVED** [review]

**Pull request closed without merging**